README.md
![Chartist Logo](https://repository-images.githubusercontent.com/320932478/2d1a7780-4ad9-11eb-809e-fbcd0458cded)

Chartist
========

Commandline tool to generate images from charsets and texts.

While [pixeling](http://grafx2.chez.com/) a [font](https://github.com/ianhan/BitmapFonts) (also known as *charset*) for a [demoscene](https://en.wikipedia.org/wiki/Demoscene) production, it would be nice to know how it will look, using the final text.

It's very common, that some letters looking good in the font itself, but not when used in a real text, consisting of different words and phrases.

To check, if the letters of the font work nicely together, you can cut'n'paste them one by one to form a complete textpage, which can take hours. Or you can use a tool to do the same job in a few seconds: **Chartist**

Chartist reads your graphicsfile with the letters (the font/charset), your textfile with all those words/phrases and will output a new graphics file, showing your text by using your pixeled letters, simple as that.

Information
========
Pixeled fonts are often used in demos for [Retrocomputers](https://en.wikipedia.org/wiki/Retrocomputing) like [Amiga](https://en.wikipedia.org/wiki/Amiga), [Atari ST](https://en.wikipedia.org/wiki/Atari_ST), [Commodore 64](https://en.wikipedia.org/wiki/Commodore_64), etc.

From [Wikipedia](https://en.wikipedia.org/wiki/Demoscene): 

> The demoscene is a computer art subculture that specializes in producing demos, which are audio-visual presentations that run in real-time on a computer. The main goal of a demo is to show off programming, artistic, and musical skills.

Those fonts are often starting with the **SPACE**-Symbol (*Pos. 32*) and ending with the **Z**-Letter (*Pos. 90*) of the [ASCII-Table](https://en.wikipedia.org/wiki/ASCII#Character_set).
If your font is not ordered this *ascii-relative* way, you can simply create a [mapping table](https://github.com/PhilSwiss/chartist/blob/master/font-mappingtable.txt) with the correct order.
Chartist supports fonts with chars from Pos. 32 (Space) up to Pos. 126 (Tilde-Symbol), so that all upper- and lowercase letters, numbers and symbols can be used.

Most of those fonts have a fixed size, e.g. **8** by **8** pixels, **16** by **16** pixels and so on. But some symbols like the **.** (dot) or the **,** (comma) look a bit lost when placed in such a grid.
To avoid this, you can create a [width table](https://github.com/PhilSwiss/chartist/blob/master/font-widthtable.txt) to define letters or symbols with a smaller width.

The **background color** is simply *auto-detected* by looking which color is used by the majority of the pixels in the font-image. But you can set it manually, if the wrong color is choosen.<br>
The amount of colors for the font is **not** limited. You can have [indexed colors](https://en.wikipedia.org/wiki/Indexed_color) aswell as [true color](https://en.wikipedia.org/wiki/Color_depth#True_color_(24-bit)),
if the are supported by the image format, for example [PNG](https://en.wikipedia.org/wiki/Portable_Network_Graphics).

The **size** of the letters is *auto-detected* by the height (*y-axis*) of the font-image. If it does not work, you can also set this value manually, expecially when the letters are arranged in multiple lines instead of one, single line.


Requirements
=============

- Python (3.5, 3.6, 3.7, 3.8, 3.9) - https://www.python.org/
- Pillow (Python Imaging Library) - https://pypi.org/project/Pillow/


Installation
=============
If missing, **Pillow** can be installed using **pip**.

Linux: 

    $ pip install pillow
Windows:

    $ python -m pip install pillow


Quickstart
==========

**Chartist** is easy to use and really straightforward to generate a fontpage.
Just run `chartist` using the provided font and the test text:

    $ chartist.py font-oneliner-asciirelative.png testtext.txt

    The generated image should be displayed by the default imageviewer of the operating system.

Write the generated image to disk:

    $ chartist.py font-oneliner-asciirelative.png testtext.txt -o fontpage.gif
	
	The submited filetype ".gif" sets the format of the output file.

Force resolution of generated image:

    $ chartist.py font-oneliner-asciirelative.png testtext.txt -r 320 256

    Otherwise the resolution is calculated by the size of the chars and the amount of text.

Render a non ascii-relative font by using a mapping table:

    $ chartist.py font-oneliner-nonasciirelative.png testtext.txt -m font-mappingtable.txt

    Because the order of the font is assumed as ascii-relative by default.

You can also use fonts which are spread over multiple lines:

    $ chartist.py font-multiline-asciirelative.png testtext.txt -s 16 16

    In this case, you have to submit the size of the chars, because autodetection will not work.

Render a proportional font (some letters have different widths) by using a width table:

    $ chartist.py font-oneliner-asciirelative-flexwidth.png testtext.txt -w font-widthtable.txt

    By default the widths of the chars/letters are of a fixed size.


Commandline options
===================

    $ chartist.py --help

    Usage: chartist charsetfile textfile [OPTION...]
    Generate images from charsets and texts
    
    mandatory arguments:
       charsetfile        imagefile with chars (e.g.: .png, .jpg, etc.)
       textfile           textfile with textline(s) (plain ascii e.g.: .txt)

    optional arguments:
       -s, --size         size of the chars (x or x and y)
	   -w, --widthtable   textfile with widthtable if chars are not of the same width
	   -m, --mappingtable textfile with mappingtable if charset is not ascii-relative
       -o, --output       outputfile with rendered text (e.g.: .png, .jpg, etc.)
       -f, --force        force overwrite of outputfile when it already exist
       -r, --resolution   resolution of imagefile with rendered text (x or x and y)
       -c, --color        backgroundcolor of imagefile with rendered text (R G B)
       -v, --version      show version info
       -h, --help         show this help

    The optional arguments are only needed if autodetection of size, resolution or
    color doesnt meet the required needs. The rendered image will only be saved, if
    an outputfile (-o/--output) is set. Otherwise the image will be shown by the os.
	The widthtable is just a textfile containing the letters, symbols, etc. and their
    individual widths one per line and seperated by a tab.
    The mappingtable is just a textfile containing all letters, symbols, etc. in the
    same order as they are positioned in the imagefile.

    examples:
       chartist charsetfile.png textfile.txt
       chartist chars.gif text.txt -o screen.png
       chartist font.tif scroll.txt -s 8 -r 320 256
       chartist letters.tif credits.txt -s 16 32 -r 256
       chartist fontset.gif phrases.txt -o words.bmp -f
       chartist graphic.jpg greets.txt -c 255 127 64 -o out.jpg
       chartist charset.png textfile.txt -s 16 38 -l mappingtable.txt
       chartist font.png textfile.txt -s 16 16 -w widthtable.txt

Files
=====

* **chartist.py** (the commandlinetool itself)
* **font-oneliner-asciirelative.png** (a simple pixeled font/charset in ascii-relative order)
* **font-oneliner-nonrelative.png** (the same font/charset in a non ascii-relative order, see mappingtable)
* **font-oneliner-asciirelative-flexwidth.png** (the same font/charset with flexible/proportional widths, see widthtable)
* **font-multiline-asciirelative.png** (the same font/charset spread over multiple lines)
* **font-mappingtable.txt** (example of a mappingtable for the non ascii-relative font/charset)
* **font-widthtable.txt** (example of a widthtable for the flexwidth font/charset)


Future ideas
============

* Output of files for the [TIC-80](https://tic80.com/) - **Done**, I have written [TicMcTile](https://github.com/PhilSwiss/ticmctile) for this job
* Output of the notorious Atari ST Scrollerformat (I only heard rumors about this one)


Bug tracker
===========

If you have any suggestions, bug reports or annoyances please report them to the issue tracker at https://github.com/PhilSwiss/chartist/issues


Contributing
============

Development of `chartist` happens at GitHub: https://github.com/PhilSwiss/chartist


Changelog
2020-05-25  Philipp
		* Version 1.0
		* Initial release

2020-06-02  Philipp
		* Version 1.1
		* Added mapping table

2020-06-04  Philipp
		* Version 1.2
		* Catch error if char is not found
		
2020-06-15  Philipp
		* Version 1.3
		* Simplified the mainloop

2020-06-24  Philipp
		* Version 1.4
                * Added width table
                * Simplified mapping table routine
		
2020-08-11  Philipp
		* Version 1.5
		* Cleaned the code with Pylama

2021-03-17  Philipp
		* Version 1.6
		* Image has same amount of colors as font, not truecolor, if possible

2021-04-13  Philipp
		* Version 1.7
		* Outputfile will now only be overwritten when forced

2022-03-15  Philipp
		* Version 1.8
		* Widthtable can now have an entry for the Space/Whitespace-character
		* Missing chars will not longer result in artifacts/empty blobs


chartist.py
#!/usr/bin/python3
#
# Chartist:
#
# a small tool to generate textscreen-images from bitmap-charsets
#
# Hint: python -m pip install pillow (install PIL on Windows)
#
# last updated by Decca / RiFT on 15.03.2022 19:05
#

# import modules
from PIL import Image
import argparse
import os.path
import sys


# replace argparse-error message with own, nicer help/usage
class ArgumentParser(argparse.ArgumentParser):
    def error(self, message):
        print("Usage: chartist charsetfile textfile [OPTION...]\n"
              "Generate images from charsets and texts\n"
              "\n"
              "mandatory arguments:\n"
              "  charsetfile        imagefile with chars (e.g.: .png, .jpg, etc.)\n"
              "  textfile           textfile with textline(s) (plain ascii e.g.: .txt)\n"
              "\n"
              "optional arguments:\n"
              "  -s, --size         size of the chars (x or x and y)\n"
              "  -w, --widthtable   textfile with widthtable if chars are not of the same width\n"
              "  -m, --mappingtable textfile with mappingtable if charset is not ascii-relative\n"
              "  -o, --output       outputfile with rendered text (e.g.: .png, .jpg, etc.)\n"
              "  -f, --force        force overwrite of outputfile when it already exist \n"
              "  -r, --resolution   resolution of imagefile with rendered text (x or x and y)\n"
              "  -c, --color        backgroundcolor of imagefile with rendered text (R G B)\n"
              "  -v, --version      show version info\n"
              "  -h, --help         show this help\n"
              "\n"
              "The optional arguments are only needed if autodetection of size, resolution or\n"
              "color does not meet the required needs. The rendered image will only be saved, if\n"
              "an outputfile (-o/--output) is set. Otherwise the image will be shown by the os.\n"
              "The widthtable is just a textfile containing the letters, symbols, etc. and their\n"
              "individual widths one per line and seperated by a tab.\n"
              "The mappingtable is also a textfile containing all letters, symbols, etc. in the\n"
              "same order as they are positioned in the imagefile.\n"
              "\n"
              "examples:\n"
              "  chartist charsetfile.png textfile.txt\n"
              "  chartist chars.gif text.txt -o screen.png\n"
              "  chartist font.tif scroll.txt -s 8 -r 320 256\n"
              "  chartist letters.tif credits.txt -s 16 32 -r 256\n"
              "  chartist fontset.gif phrases.txt -o words.bmp -f\n"
              "  chartist graphic.jpg greets.txt -c 255 127 64 -o out.jpg\n"
              "  chartist charset.png textfile.txt -s 16 38 -m mappingtable.txt\n"
              "  chartist font.png textfile.txt -s 16 16 -w widthtable.txt\n", file=sys.stderr)
        self.exit(1, '%s: ERROR: %s\n' % (self.prog, message))


# set commandline arguments
parser = ArgumentParser(prog='chartist', add_help=False)
parser.add_argument('charset',
                    metavar='charsetfile.png',
                    type=str,
                    help='imagefile with chars')
parser.add_argument('text',
                    metavar='textfile.txt',
                    type=str,
                    help='textfile with textline(s)')
parser.add_argument('-s', '--size',
                    metavar='8',
                    type=int,
                    nargs='+',
                    action='store',
                    help='size of the charset (x or x and y)')
parser.add_argument('-w', '--widthtable',
                    metavar='widthtable.txt',
                    type=str,
                    action='store',
                    help='textfile with width-table')
parser.add_argument('-m', '--mappingtable',
                    metavar='mappingtable.txt',
                    type=str,
                    action='store',
                    help='textfile with mapping-table')
parser.add_argument('-o', '--output',
                    metavar='output.png',
                    type=str,
                    action='store',
                    help='imagefile with rendered text')
parser.add_argument('-f', '--force',
                    action='store_true',
                    help='force overwrite of imagefile')
parser.add_argument('-r', '--resolution',
                    metavar='320',
                    type=int,
                    nargs='+',
                    action='store',
                    help='resolution of imagefile with rendered text  (x or x and y)')
parser.add_argument('-c', '--color',
                    metavar='255',
                    type=int,
                    nargs=3,
                    action='store',
                    help='maincolor of imagefile with rendered text (R G B)')
parser.add_argument('-v', '--version',
                    action='version',
                    version='%(prog)s 1.8')
args = parser.parse_args()


# check size argument
if args.size is not None and len(args.size) not in (1, 2):
    parser.error("either give 1 or 2 values for size, not " + str(len(args.size)))


# check resolution argument
if args.resolution is not None and len(args.resolution) not in (1, 2):
    parser.error("either give 1 or 2 values for resolution, not " + str(len(args.resolution)))


# init vars
offsetX = 0
offsetY = 0
locationX = 0
locationY = 0


# get commandline arguments
imageFile = args.charset
textFile = args.text
charArgs = args.size
widthFile = args.widthtable
mappingFile = args.mappingtable
outputFile = args.output
outputForce = args.force
outputRes = args.resolution
colorArgs = args.color


# load image file
try:
    orgImg = Image.open(imageFile)
except Exception as error:
    print("ERROR: " + str(error), file=sys.stderr)
    exit(1)
print("    Charset: " + imageFile)


# get image mode & format
orgMode = orgImg.mode
orgFormat = orgImg.format
orgFormat = '.' + orgFormat.lower()
print("     Format: " + orgFormat + " (" + orgMode + ")")


# get image dimensions
orgSizeX, orgSizeY = orgImg.size
print(" Resolution: " + str(orgSizeX) + " x " + str(orgSizeY))


# get image colors & amount
orgColors = orgImg.convert('RGB').getcolors(maxcolors=(orgSizeX * orgSizeY))
print("     Colors: " + str(len(orgColors)))


# load text file
textLines = []
try:
    with open(textFile) as file:
        for line in file:
            textLines.append(line.rstrip('\n'))
except Exception as error:
    print("ERROR: " + str(error), file=sys.stderr)
    exit(1)


# get amount of textlines
amountLines = len(textLines)
# get lenght of longest textline
longestLine = len(max(textLines, key=len))
print("  Textlines: " + str(amountLines) + " (max. " + str(longestLine) + " chars)")


# get or detect size of chars
if isinstance(charArgs, list):
    charSizeX = charArgs[0]
    if len(charArgs) > 1:
        charSizeY = charArgs[1]
    else:
        charSizeY = charArgs[0]
    print("   Charsize: " + str(charSizeX) + " x " + str(charSizeY))
else:
    if orgSizeX < orgSizeY:
        charSizeX = orgSizeX
        charSizeY = orgSizeX
    else:
        charSizeX = orgSizeY
        charSizeY = orgSizeY
    print("   Charsize: " + str(charSizeX) + " x " + str(charSizeY) + " (auto)")


# get or detect output resolution
if isinstance(outputRes, list):
    newImgX = outputRes[0]
    if len(outputRes) > 1:
        newImgY = outputRes[1]
    else:
        newImgY = outputRes[0]
    print("    Preview: " + str(newImgX) + " x " + str(newImgY))
else:
    # width by longest textline
    newImgX = longestLine * charSizeX
    # height by amount of textlines
    newImgY = amountLines * charSizeY
    print("    Preview: " + str(newImgX) + " x " + str(newImgY) + " (auto)")


# get or generate width-table
widthTable = {}
if isinstance(widthFile, str):
    try:
        with open(widthFile) as file:
            for line in file:
                letter, width = line.split('\t', 1)
                widthTable[letter] = int(width.strip())
    except Exception as error:
        print("ERROR: " + str(error), file=sys.stderr)
        exit(1)
    print("     Widths: " + widthFile + " (" + str(len(widthTable)) + " entries)")
    # fill up width-table if incomplete
    if len(widthTable) < 96:
        for asc in range(32, 128):
            if not chr(asc) in widthTable:
                widthTable[chr(asc)] = charSizeX
        print("     Widths: added missing entries")
else:
    for asc in range(32, 128):
        widthTable[chr(asc)] = charSizeX
    print("     Widths: " + str(charSizeX) + " pixels (auto)")


# get or generate mapping-table
mappingTable = []
if isinstance(mappingFile, str):
    try:
        with open(mappingFile) as file:
            for line in file:
                for char in line.rstrip('\n'):
                    mappingTable.append(char)
    except Exception as error:
        print("ERROR: " + str(error), file=sys.stderr)
        exit(1)
    print("    Mapping: " + mappingFile + " (" + str(len(mappingTable)) + " entries)")
    # fill up mapping-table if incomplete
    if len(mappingTable) < 96:
        for asc in range(32, 128):
            if not chr(asc) in mappingTable:
                mappingTable.append(chr(asc))
        print("    Mapping: added missing chars")
else:
    for asc in range(32, 128):
        mappingTable.append(chr(asc))
    print("    Mapping: ascii-relative (auto)")


# get or detect background color
if isinstance(colorArgs, list):
    backgroundColor = tuple(colorArgs)
    print("    BGcolor: " + str(backgroundColor).replace("(", "").replace(")", ""))
else:
    # get dominant color
    dominantColor = max(orgColors)
    backgroundColor = dominantColor[1]
    print("    BGcolor: " + str(backgroundColor).replace("(", "").replace(")", " (auto)"))


# create new, empty image, but keep mode, colors, palette, etc.
if len(orgMode) > 1:
    # if mode rgb or higher, just set background via rgb-value
    newImg = orgImg.resize((newImgX, newImgY))
    newImg.paste(backgroundColor, (0, 0, newImgX, newImgY))
else:
    # if mode palette or lower, try to map rgb-value...
    pixelsRGB = list(orgImg.convert('RGB').getdata())
    indexRGB = next((index for index, value in enumerate(pixelsRGB) if value == backgroundColor), None)
    # ...to the existing palette
    if indexRGB is not None:
        pixelsPalette = list(orgImg.getdata())
        indexPalette = pixelsPalette[indexRGB]
        newImg = orgImg.resize((newImgX, newImgY))
        newImg.paste(indexPalette, (0, 0, newImgX, newImgY))
    # or create new rgb-image if no match is found
    else:
        newImg = Image.new(mode="RGB", size=(newImgX, newImgY), color=backgroundColor)


# generate text with chars
print("    generating image...")
for line in textLines:
    for letter in line:
        # get char position from mapping-table
        offsetX = (mappingTable.index(letter)) * charSizeX
        # if charset-image is NOT a one-liner, get chars from multiple lines
        if offsetX >= orgSizeX:
            charLine = (offsetX // orgSizeX)
            offsetY = (charSizeY * charLine)
            offsetX = offsetX - (orgSizeX * charLine)
            # if offset is out of bounds, the char was not found
            if offsetY >= orgSizeY:
                print("    ERROR: unmatched char \"" + letter + "\"", file=sys.stderr)
        else:
            offsetY = 0
        # copy char, but only from inbound charset-image
        Tile = orgImg.crop((offsetX, offsetY, min(orgSizeX, (offsetX + charSizeX)), min(orgSizeY, (offsetY + charSizeY))))
        # paste char to new image
        newImg.paste(Tile, (locationX, locationY))
        # count to the next pasting x-position by width of char
        locationX = locationX + widthTable[letter]
    # count to next pasting y-position
    locationY = locationY + charSizeY
    locationX = 0


# show final image when no outputfile
if not isinstance(outputFile, str):
    print("    try to show image...")
    newImg.show()
    print("    done.")
    exit()


# check if output file already exist
def check_file(outputName):
    if not outputForce:
        if os.path.isfile(outputName):
            print("ERROR: file already exist")
            exit(1)
    else:
        return


# save final image to outputfile
print("    try to save: " + str(outputFile))
check_file(outputFile)
try:
    newImg.save(outputFile)
except ValueError as error:
    if 'unknown file extension' in str(error):
        print("    try to save: " + str(outputFile) + str(orgFormat))
        check_file(outputFile + orgFormat)
        try:
            newImg.save(outputFile + orgFormat)
        except Exception as error:
            print("ERROR: " + str(error), file=sys.stderr)
            exit(1)
    else:
        print("ERROR: " + str(error), file=sys.stderr)
        exit(1)
except Exception as error:
    print("ERROR: " + str(error), file=sys.stderr)
    exit(1)


# end message
print("    done.")


# end of code
exit()
